scopes/webpack/webpack/config/webpack.dev.config.ts

```diff
@@ -109,30 +109,29 @@ export function configFactory(
         overlay: false,
       },
 
-      onBeforeSetupMiddleware(wds) {
-        const { app } = wds;
+      setupMiddlewares: (middlewares, devServer) => {
+        if (!devServer) {
+          throw new Error('webpack-dev-server is not defined');
+        }
+
         // Keep `evalSourceMapMiddleware` and `errorOverlayMiddleware`
         // middlewares before `redirectServedPath` otherwise will not have any effect
         // This lets us fetch source contents from webpack for the error overlay
-        // @ts-ignore - @types/WDS mismatch - 3.11.6 vs 4.0.3
-        app.use(evalSourceMapMiddleware(wds));
-        // This lets us open files from the runtime error overlay.
-        // @ts-ignore
-        app.use(errorOverlayMiddleware());
-      },
-
-      onAfterSetupMiddleware({ app }) {
-        // Redirect to `PUBLIC_URL` or `homepage` from `package.json` if url not match
-        // @ts-ignore
-        app.use(redirectServedPath(publicUrlOrPath));
-
-        // This service worker file is effectively a 'no-op' that will reset any
-        // previous service worker registered for the same host:port combination.
-        // We do this in development to avoid hitting the production cache if
-        // it used the same host and port.
-        // https://github.com/facebook/create-react-app/issues/2272#issuecomment-302832432
-        // @ts-ignore
-        app.use(noopServiceWorkerMiddleware(publicUrlOrPath));
+        middlewares.push(
```

Comment:
you should not push `onBeforeSetupMiddleware`, you should use `unshift` instead:
```ts
      setupMiddlewares(middlewares: any, devServer: any) {
        // Keep `evalSourceMapMiddleware` and `errorOverlayMiddleware`
        // middlewares before `redirectServedPath` otherwise will not have any effect
        // This lets us fetch source contents from webpack for the error overlay
        middlewares.unshift(evalSourceMapMiddleware(devServer));
        // This lets us open files from the runtime error overlay.
        middlewares.unshift(errorOverlayMiddleware());

        // Redirect to `PUBLIC_URL` or `homepage` from `package.json` if url not match
        middlewares.push(redirectServedPath(publicUrlOrPath));
        // This service worker file is effectively a 'no-op' that will reset any
        // previous service worker registered for the same host:port combination.
        // We do this in development to avoid hitting the production cache if
        // it used the same host and port.
        // https://github.com/facebook/create-react-app/issues/2272#issuecomment-302832432
        middlewares.push(noopServiceWorkerMiddleware(publicUrlOrPath));

        return middlewares;
      },
```

Reply:
This is not really needed in that case.
it's only needs to be before the `redirectServedPath` (see the comments above)

Reply:
ok